### PR TITLE
feat: insert delete rows cols

### PR DESCRIPTION
**Pull request by BrianHung**
bulk insert delete rows cols

### CODE REVIEW

**Review by nhatcher — CHANGES_REQUESTED** [review]
This is great!

Some comments:

1. we shouldn't keep the old API for insert 1 row or column
2. We need tests in the Rust code (its also fine to have them in JavaScript. Test should cover corner cases, like too many columns, negative values, etc

Point 1. will imply changing the UI a little bit. You don't need to add UI to insert bulk columns/rows as @dg-ac will need to do some design, juts adapt the API call.

It was a mistake to do it like this for me in the first place. Can't wait for this to be in place.

**Comment by BrianHung** [comment]
I didn't know whether to remove the singular case to keep things backwards compatible or not. Might require a version bump if we do? If you're fine with it, I'll remove it!

**Comment by nhatcher** [comment]
> I didn't know whether to remove the singular case to keep things backwards compatible or not. Might require a version bump if we do? If you're fine with it, I'll remove it!

Yes, please remove them. I'll handle releases and version bumps. You will need to update bindings and the js code of course, but should be pretty straightforward. you will need to change places like this:
https://github.com/ironcalc/IronCalc/blob/main/webapp/IronCalc/src/components/Worksheet/Worksheet.tsx#L343
But you don't need to update the UI or anything

**Comment by BrianHung** [comment]
Do you think we should also support `InsertRows` `InsertCols` `DeleteRows` `DeleteCols` as diff types? Would be easier to multiplayer support if it builds on top of diffs 🤔 Can do in follow-up PR.

File: base/src/user_model/history.rs [hunk 1]
```diff
@@ -1,3 +1,5 @@
+#![allow(deprecated)]
```
Comment (nhatcher) [inline base/src/user_model/history.rs]: Please remove all deprecated functions. We really don't need them. At this point it is ok to break this.

File: base/src/user_model/common.rs [hunk 2]
```diff
@@ -2478,4 +2614,6 @@ mod tests {
             assert_eq!(horizontal(&format!("{a}")), Ok(a));
         }
     }
+
+    // (batch diff tests moved to separate file)
```
Comment (nhatcher) [inline base/src/user_model/common.rs]: You can remove this two lines

File: base/src/test/user_model/test_batch_row_column_diff.rs [hunk 3]
```diff
@@ -0,0 +1,681 @@
...
+    let list = last_diff_list(&mut model);
+    assert_eq!(list.len(), 1);
+    match &list[0] {
+        Diff::InsertRows { sheet, row, count } => {
+            assert_eq!(*sheet, 0);
+            assert_eq!(*row, 5);
+            assert_eq!(*count, 3);
+        }
+        _ => panic!("Unexpected diff variant"),
+    }
```
Comment (nhatcher) [inline base/src/test/user_model/test_batch_row_column_diff.rs]: A slightly nicer way to do this is:
```
assert!(matches!(list[0], Diff::InsertRows { sheet: 0, row: 5, count: 3 }));
```

**Review by nhatcher — CHANGES_REQUESTED** [review]
Almost there!
Please remove the deprecated methods. Once taht's done I'll approve.

File: base/src/user_model/history.rs (hunk 1 above)
Comment (BrianHung) [reply base/src/user_model/history.rs]: Will do!

**Comment by BrianHung** [comment]
wasm-pack / CI CD broken upstream but tests + code works locally for me.

**Review by nhatcher — APPROVED** [review]
LGTM!

**Comment by nhatcher** [comment]
Could you rebase main? There is a small conflict, but easy to resolve

Many thanks for this feature!

**Comment by BrianHung** [comment]
pushed force with lease; would just squash and merge all the commits!
